src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,365 @@
+ * labels. The output is in the form of a matrix, which leads to simpler
+ * calculations in the Evaluate() and Gradient() methods.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::GetGroundTruthMatrix(
+        const arma::Row<size_t>& labels, arma::sp_mat& groundTruth)
+{
+  // Calculate the ground truth matrix according to the labels passed. The
+  // ground truth matrix is a matrix of dimensions 'numClasses * numExamples',
+  // where each column contains a single entry of '1', marking the label
+  // corresponding to that example.
+
+  // Row pointers and column pointers corresponding to the entries.
+  arma::uvec rowPointers(labels.n_elem);
+  arma::uvec colPointers(labels.n_elem + 1);
+
+  // Row pointers are the labels of the examples, and column pointers are the
+  // number of cumulative entries made uptil that column.
+  for (size_t i = 0; i < labels.n_elem; i++)
+  {
+    rowPointers(i) = labels(i);
+    colPointers(i + 1) = i + 1;
+  }
+
+  // All entries are '1'.
+  arma::vec values;
+  values.ones(labels.n_elem);
+
+  // Calculate the matrix.
+  groundTruth = arma::sp_mat(rowPointers, colPointers, values, numClasses,
+                             labels.n_elem);
+}
+
+/**
+ * Shuffle the data.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::Shuffle()
+{
+  // Determine new ordering.
+  arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
+      dataset.n_cols - 1, dataset.n_cols));
+
+  // Re-sort data.
+  arma::mat newData = dataset.cols(ordering);
+  math::ClearAlias(dataset);
+  dataset = std::move(newData);
+
+  // Assemble data for batch constructor.  We need reverse orderings though...
+  arma::uvec reverseOrdering(ordering.n_elem);
+  for (size_t i = 0; i < ordering.n_elem; ++i)
+    reverseOrdering[ordering[i]] = i;
+
+  arma::umat newLocations(2, groundTruth.n_nonzero);
+  arma::vec values(groundTruth.n_nonzero);
+  arma::sp_mat::const_iterator it = groundTruth.begin();
+  size_t loc = 0;
+  while (it != groundTruth.end())
+  {
+    newLocations(0, loc) = reverseOrdering(it.col());
+    newLocations(1, loc) = it.row();
+    values(loc) = (*it);
+
+    ++it;
+    ++loc;
+  }
+
+  groundTruth = arma::sp_mat(newLocations, values, groundTruth.n_rows,
+                             groundTruth.n_cols);
+}
+
+template <typename MatType>
+double LinearSVMFunction<MatType>::Evaluate(
+    const arma::mat& parameters)
+{
+  // The objective function is the hinge loss function and it is
+  // calculated over all the training examples.
+
+  // Calculate the loss and regularization terms.
+  double loss, regularization, cost;
+
+  arma::mat scores = parameters * dataset;
+  arma::sp_mat correctScores = scores % groundTruth;
+  arma::mat margin = scores
+      - arma::repmat(arma::ones(numClasses).t() * correctScores, numClasses, 1)
+      + 1 - groundTruth;
+
+  // The Hinge Loss Function
+  loss = arma::accu(arma::clamp(margin, 0.0, margin.max()));
```

Comment:
You could just use `DBL_MAX` or similar for the upper bound; it'll take non-constant time to compute `margin.max()`.